stratosphere/ams_mitm/source/utils.cpp

```diff
@@ -665,3 +667,91 @@ Result Utils::GetSettingsItemBooleanValue(const char *name, const char *key, boo
 void Utils::RebootToFatalError(AtmosphereFatalErrorContext *ctx) {
     BpcRebootManager::RebootForFatalError(ctx);
 }
+
+void Utils::CreateBlankProdinfoIfNeeded() {
+    const bool p = Utils::GetCAL0BackupBufferToBlank();
+    
+    Result rc;
+    FsFile file;
+
+    rc = fsFsOpenFile(&g_sd_filesystem, "/atmosphere/automatic_backups/prodinfo_blank.bin", FS_OPEN_READ, &file);
+    if (R_SUCCEEDED(rc)) {
+        fsFileClose(&file);
+        
+        return;
+    }
+
+    // big brain: modify buffer inline
+
+    const char serial[] = "XAW00000000000";
+
+    memcpy(p ? &g_cal0_backup[0x250] : &g_cal0_storage_backup[0x250], serial, sizeof(serial) - 1);
+
+    static constexpr u64 erase_offsets[7] = {0xAE0, 0x3AE0, 0x35E1, 0x36E1, 0x2B0, 0x3D70, 0x3FC0};
+    static constexpr u64 erase_sizes[7] = {0x800, 0x130, 0x6, 0x6, 0x180, 0x240, 0x240};
+
+    for (size_t i = 0; i < 7; i++) {
+        u64 offset = erase_offsets[i];
+        u64 size = erase_sizes[i];
+
+        memset(p ? &g_cal0_backup[offset] : &g_cal0_storage_backup[offset], 0, size);
+    }
+
+    static constexpr u64 hash_offsets[2] = {0x12E0, 0x20};
+    static constexpr u64 data_offsets[2] = {0xAE0, 0x40};
+    u64 data_sizes[2] = {p ? g_cal0_backup[0xAD0] : g_cal0_storage_backup[0xAD0], p ? g_cal0_backup[0x8] : g_cal0_storage_backup[0x8]};
+
+    for (size_t i = 0; i < 2; i++) {
+        u64 hash_offset = hash_offsets[i];
+        u64 data_offset = data_offsets[i];
+        u64 data_size = data_sizes[i];
+
+        u8 data[data_size];
+
+        memcpy(&data, p ? &g_cal0_backup[data_offset] : &g_cal0_storage_backup[data_offset], data_size);
+
+        u8 hash[0x20];
+
+        Sha256Context ctx;
+        sha256ContextCreate(&ctx);
+
+        sha256ContextUpdate(&ctx, data, data_size);
+
+        sha256ContextGetHash(&ctx, &hash);
+
+        memcpy(p ? &g_cal0_backup[hash_offset] : &g_cal0_storage_backup[hash_offset], hash, 0x20);
+    }
+
+    rc = fsFsOpenFile(&g_sd_filesystem, "/atmosphere/automatic_backups/prodinfo_blank.bin", FS_OPEN_WRITE, &file);
+    if (R_FAILED(rc)) {
+        std::abort();
+    }
+
+    rc = fsFileSetSize(&file, ProdinfoSize);
+    if (R_FAILED(rc)) {
+        std::abort();
+    }
+
+    rc = fsFileWrite(&file, 0, p ? g_cal0_backup : g_cal0_storage_backup, ProdinfoSize);
+    if (R_FAILED(rc)) {
+        std::abort();
+    }
+
+    rc = fsFileFlush(&file);
+    if (R_FAILED(rc)) {
+        std::abort();
+    }
+
+    fsFileClose(&file);
+}
+
+Result Utils::OpenBlankProdinfoFile(FsFile *out) {
+    return fsFsOpenFile(&g_sd_filesystem, "/atmosphere/automatic_backups/prodinfo_blank.bin", FS_OPEN_READ, out);
```

Comment:
Duplicate persisted read handle instead of opening new reference.

Reply:
https://canary.discordapp.com/channels/269333940928512010/420029476634886144/572926601906356234